### PR TITLE
drivers: modem: cmd_handler: Allow locking TX

### PR DESCRIPTION
**Pull request by endian-benjamin**
Note: this depends on #28475 

PR #28475 required user to explicitly ensure there was no concurrent modem UART access before restarting gsm_ppp stack. This PR fixes this and adds a generally useful method of blocking access to modem UART via modem_cmd_send which is useful for things other than gsm_ppp.

### CODE REVIEW

**Comment by endian-benjamin** [comment]
Tested on SIMCOM7600E modem with GSM_MUX enabled, with multiple independent users of the AT channel.

File: drivers/modem/gsm_ppp.c [hunk 1]
```diff
@@ -336,12 +336,12 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
 	int ret;
 
 	if (IS_ENABLED(CONFIG_GSM_MUX) && gsm->mux_enabled) {
-		ret = modem_cmd_send(&gsm->context.iface,
+		ret = modem_cmd_send_nolock(&gsm->context.iface,
 				     &gsm->context.cmd_handler,
 				     &response_cmds[0],
```
Comment (jukkar) [inline drivers/modem/gsm_ppp.c:341]: Indentation of these two lines needs fixing

**Comment by nvlsianpu** [comment]
Please rebase on master and follow the contribution guide.

File: samples/net/gsm_modem/src/main.c [hunk 2]
```diff
@@ -6,15 +6,59 @@
...
 
+static const struct device *gsm_dev;
 static struct net_mgmt_event_callback mgmt_cb;
+static bool connected;
+
+static int cmd_sample_modem_suspend(const struct shell *shell,
+				    size_t argc, char *argv[])
+{
+	if (!connected) {
+		LOG_INF("Not connected.");
```
Comment (jakub-uC) [inline samples/net/gsm_modem/src/main.c:27]: Why not use `shell_print(shell, "Not connected.");` ?

File: samples/net/gsm_modem/src/main.c [hunk 3]
```diff
@@ -6,15 +6,59 @@
...
+	gsm_ppp_stop(gsm_dev);
+
+	return 0;
+}
+
+static int cmd_sample_modem_resume(const struct shell *shell,
+				   size_t argc, char *argv[])
+{
+	if (connected) {
+		LOG_INF("Already connected.");
```
Comment (jakub-uC) [inline samples/net/gsm_modem/src/main.c:40]: Why not use shell_print(shell, "Already connected."); ?

File: samples/net/gsm_modem/src/main.c (hunk 2 above)
Comment (jukkar) [reply samples/net/gsm_modem/src/main.c:27]: Yep, that would actually be a more proper print, I just copied this from other code thus logging print was used.
Anyway, this commit does not really need to be part of this PR, it just helps when testing/debugging the feature.
@endian-benjamin I think we need to first get #28475 merged as it and this PR contain shared commits, and then come back to this fix.

Comment (endian-benjamin) [reply samples/net/gsm_modem/src/main.c:27]: I think you're missing the fact that this is based on an unmerged PR :)

Perhaps this should be marked WIP or something until it's merged? What's the procedure?

**Comment by carlescufi** [comment]
@endian-benjamin #28475 is now merged.

**Review by jukkar — APPROVED** [review]
LGTM

**Review by rlubos — APPROVED** [review]

**Pull request merged**